### ISSUE TITLE
Behavior incorrect of basic_streambuf with concepts override

### ISSUE THREAD

**Issue by expnkx** [microsoft/STL#574]
**Describe the bug**
```cpp
template<stream stm,typename traits_tp = std::char_traits<typename stm::char_type>>
class fast_io_basic_streambuf final:public std::basic_streambuf<typename traits_tp::char_type,traits_tp>
{
public:
	using native_handle_type = stm;
	using traits_type = traits_tp;
	using char_type = typename traits_type::char_type;
private:
	native_handle_type sm;
	std::streamsize xsputn(char_type const* s, std::streamsize count) requires(output_stream<native_handle_type>)
	{
		if constexpr(!std::same_as<decltype(write(sm,s,s+count)),void>)
			return static_cast<std::streamsize>(write(sm,s,s+count)-s);
		else
		{
			write(sm,s,s+count);
			return count;
		}
	}
	std::streamsize xsgetn(char_type* s, std::streamsize count) requires(input_stream<native_handle_type>)
	{
		return static_cast<std::streamsize>(read(sm,s,s+count)-s);
	}
public:
	template<typename... Args>
	requires std::constructible_from<stm,Args...>
	fast_io_basic_streambuf(Args&& ...args):sm(std::forward<Args>(args)...){}
	constexpr auto& native_handle()
	{
		return sm;
	}
};
```
**Command-line test case**
cl -EHsc -Ox output_10M_size_t.cc /MD /std:c++latest /experimental:concepts syserror.obj

STL version (git commit or Visual Studio version):
Master

The code compiles. However, MSVC does not generate the correct output for this. GCC generates the right one with same code.
I do not know whether C++ standard said something about how to use concepts to override derived class methods.

The source code is here:
https://bitbucket.org/ejsvifq_mabmip/fast_io/src/master/include/fast_io_legacy_impl/fast_io_streambuf.h
https://bitbucket.org/ejsvifq_mabmip/fast_io/src/master/benchmarks/0000.10m_size_t/output_10M_size_t.cc

BTW: the implementation of std::basic_filebuf in MSVC is just a wrapper of FILE*.
https://bitbucket.org/ejsvifq_mabmip/fast_io/src/master/include/fast_io_legacy_impl/c_style.h
https://bitbucket.org/ejsvifq_mabmip/fast_io/src/master/include/fast_io_legacy_impl/stream_file.h

**Comment by StephanTLavavej** [comment]
This is just a class template, but we need a self-contained test case - i.e. something with `main()` that will actually compile and link. Additionally, we need to know exactly what output you're observing and expecting, so we can determine if we're reproducing the issue correctly.

**Comment by expnkx** [comment]
> This is just a class template, but we need a self-contained test case - i.e. something with `main()` that will actually compile and link. Additionally, we need to know exactly what output you're observing and expecting, so we can determine if we're reproducing the issue correctly.

This template is to redirect istream/ostream to fast_io's devices. However, it does not work correctly on MSVC. I think the override failed???

**Comment by expnkx** [comment]
This is MSVC's output
![image](https://user-images.githubusercontent.com/57303909/75618451-9f7a9e00-5b3c-11ea-8698-7f91685068a3.png)
GCC's output
![image](https://user-images.githubusercontent.com/57303909/75618468-cc2eb580-5b3c-11ea-87ed-a6b54225ba09.png)
You can see fast_io_streambuf.txt size on MSVC is not correct.

**Comment by CaseyCarter** [comment]
`fast_io_basic_streambuf::xsgetn` and `fast_io_basic_streambuf::xsputn` are virtual functions since `basic_streambuf::xsgetn` and `basic_streambuf::xsputn` are virtual functions. [[class.virtual]/6](http://eel.is/c++draft/class.virtual#6) states "A virtual function shall not have a trailing _requires-clause_." Consequently, this class template is ill-formed and it's a bug that neither MSVC nor GCC diagnose it as such (https://godbolt.org/z/rJef_V). I've reported the MSVC bug as DevCom-934979.

Does removing the *requires-clause*s from those virtual functions resolve the discrepancy you're seeing?

**Comment by expnkx** [comment]
> `fast_io_basic_streambuf::xsgetn` and `fast_io_basic_streambuf::xsputn` are virtual functions since `basic_streambuf::xsgetn` and `basic_streambuf::xsputn` are virtual functions. [[class.virtual]/6](http://eel.is/c++draft/class.virtual#6) states "A virtual function shall not have a trailing _requires-clause_." Consequently, this class template is ill-formed and it's a bug that neither MSVC nor GCC diagnose it as such (https://godbolt.org/z/rJef_V). I've reported the MSVC bug as [DevCom-934979](https://developercommunity.visualstudio.com/content/problem/934979/meow.html).
> 
> Does removing the _requires-clause_s from those virtual functions resolve the discrepancy you're seeing?

Thank you for your response. I think it is a legitimate use case to conditionally override virtual functions.

I think a defect report should be sent to ISOCPP. That is one that I found concepts are not compatible with the OOP mechanism which is a huge issue.

**Comment by CaseyCarter** [comment]
FYI, DevCom-934979 has been internally fixed for Visual Studio 2019 version 16.7, so MSVC will correctly diagnose virtual functions with constraints as ill-formed. If you'd like to suggest to WG21 that virtual functions should allow constraints I suggest you start at https://isocpp.org/std/submit-issue.

Thanks for the report!
